walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,188 @@
 
 package com.google.sps;
 
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.ListIterator;
 
 public final class FindMeetingQuery {
+  private int meetingDuration;
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    // The total (common) free time we have to conduct the meeting.
+    LinkedList<TimeRange> freeTime = new LinkedList<TimeRange>();
+    freeTime.addAll(Arrays.asList(TimeRange.WHOLE_DAY));
+
+    meetingDuration = (int) request.getDuration();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    Collection<String> mandatoryAttendees = request.getAttendees();
+
+    // If duration of the meeting is greater than a whole day, then return an empty list.
+    if (meetingDuration > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+
+    // If there are no attendees then return the whole day (ie the free time we have till now).
+    if (mandatoryAttendees.isEmpty() && optionalAttendees.isEmpty()) {
+      return freeTime;
+    }
+
+    // Changes freeTime such that it now contains all the free time slots when the meeting can
+    // happen, updating the events that optional attendees can attend.
+    freeSlotsWithMandatoryAttendees(
+        freeTime, events, mandatoryAttendees);
+
+    // If free time is not empty and optional attendees are not empty
+    // then go ahead and look for time slots to have meeting with our optional attendees.
+    if (!freeTime.isEmpty()) {
+      freeTime =
+          freeSlotsWithOptionalAttendees(events, freeTime, optionalAttendees);
+    }
+
+    return freeTime;
+  }
+
+  /** Loop through all the events given and modify freeTime List.
+   */
+  private void freeSlotsWithMandatoryAttendees(
+      LinkedList<TimeRange> freeTime,
+      Collection<Event> events,
+      Collection<String> mandatoryAttendees) {
+
+    for (Event event : events) {
+      // Check if the event has any mandatory attendees and 
+      // then sends it to find conflict and reslove method
+      if(!Collections.disjoint(mandatoryAttendees, event.getAttendees())) {
+        findConflictAndResolve(freeTime, event);
+      }
+    }
+
+  }
+
+  /** Try checking for any time slots we can get for the meeting with optional attendees.
+   */
```

Comment:
For JavaDoc it's common to start with `/**` on a separate line. So this should be:
```
  /**
   * Try checking for any time slots we can get for the meeting with optional attendees.
   */
```